rs/replicated_state/src/replicated_state.rs

```diff
@@ -194,15 +195,19 @@ impl std::iter::Iterator for OutputIterator<'_> {
     /// for that canister, the canister iterator is moved to the back of the
     /// iteration order.
     fn next(&mut self) -> Option<Self::Item> {
-        if let Some(mut canister_iterator) = self.canister_iterators.pop_front() {
-            if let Some(msg) = canister_iterator.next() {
-                self.size -= 1;
+        while let Some(mut canister_iterator) = self.canister_iterators.pop_front() {
```

Comment:
Is this just defensive programming or could it really happen given our invariants that an entire `canister_iterator` is empty and we enter this loop a second time?

Reply:
Right now it's just accidental defensive programming. We only create a `CanisterOutputQueuesIterator` for canisters whose `CanisterQueues::has_output()` returns `true`, i.e. there is an actual outbound message in the pool.

I changed this when I had dropped the "non-stale reference at the front of each queue" invariant; and then forgot to undo the change when I reintroduced it.

I switched it back to an `if` and, as expected, everything works, including proptests that produce canisters with empty output queues. Thanks for the catch.